op-program/host/host.go

```diff
@@ -51,83 +58,193 @@ func FaultProofProgram(logger log.Logger, cfg *config.Config) error {
+			l2ClCfg := sources.L2ClientDefaultConfig(cfg.Rollup, true)
+			l1Cl, err := sources.NewL1Client(l1RPC, logger, nil, l1ClCfg)
+			if err != nil {
+				return fmt.Errorf("failed to create L1 client: %w", err)
+			}
+			l2Cl, err := sources.NewL2Client(l2RPC, logger, nil, l2ClCfg)
+			if err != nil {
+				return fmt.Errorf("failed to create L2 client: %w", err)
+			}
+			l2DebugCl := &L2Source{L2Client: l2Cl, DebugClient: sources.NewDebugClient(l2RPC.CallContext)}
+
+			logger.Info("Setting up pre-fetcher")
+			prefetch := prefetcher.NewPrefetcher(l1Cl, l2DebugCl, kv)
+			getPreimage = func(key common.Hash) ([]byte, error) { return prefetch.GetPreimage(ctx, key) }
+			hinter = prefetch.Hint
+		} else {
+			logger.Info("Using offline mode. All required pre-images must be pre-populated.")
+			getPreimage = kv.Get
+			hinter = func(hint string) error {
+				logger.Debug("ignoring prefetch hint", "hint", hint)
+				return nil
+			}
 		}
+	}
 
-		logger.Info("Connecting to L2 node", "l2", cfg.L2URL)
-		l2RPC, err := client.NewRPC(ctx, logger, cfg.L2URL)
+	// Setup client I/O for preimage oracle interaction
+	var pClientRW *readWritePair
+	if IsHost() {
+		var pHostRW *readWritePair
+		var err error
+		pClientRW, pHostRW, err = bidirectionalPipe()
 		if err != nil {
-			return fmt.Errorf("failed to setup L2 RPC: %w", err)
+			return fmt.Errorf("failed to create preimage pipe: %w", err)
 		}
+		oracleServer := preimage.NewOracleServer(pHostRW)
+		launchOracleServer(logger, oracleServer, getPreimage)
+		defer pHostRW.Close()
+	} else {
+		r := os.NewFile(pClientRFd, "pClientR")
+		w := os.NewFile(pClientWFd, "pClientW")
+		pClientRW = &readWritePair{r, w}
+	}
 
-		l1ClCfg := sources.L1ClientDefaultConfig(cfg.Rollup, cfg.L1TrustRPC, cfg.L1RPCKind)
-		l2ClCfg := sources.L2ClientDefaultConfig(cfg.Rollup, true)
-		l1Cl, err := sources.NewL1Client(l1RPC, logger, nil, l1ClCfg)
+	// Setup client I/O for hint comms
+	var hClientW *os.File
+	if IsHost() {
+		var err error
+		var hHostR *os.File
+		hHostR, hClientW, err = os.Pipe()
 		if err != nil {
-			return fmt.Errorf("failed to create L1 client: %w", err)
+			return fmt.Errorf("failed to create hint pipe: %w", err)
+		}
+		defer hHostR.Close()
+		hHost := preimage.NewHintReader(hHostR)
+		routeHints(logger, hHost, hinter)
+	} else {
+		hClientW = os.NewFile(hClientWFd, "hClientW")
+	}
+
+	var childPID int
+	if IsHost() && cfg.Detached {
+		env := os.Environ()
+		env = append(env, "OP_PROGRAM_CHILD=true")
+		files := make([]uintptr, maxFd)
+		files[0] = os.Stdin.Fd()
+		files[1] = os.Stdout.Fd()
+		files[2] = os.Stderr.Fd()
+		files[hClientWFd] = hClientW.Fd()
+		files[pClientRFd] = pClientRW.ReadFd()
+		files[pClientWFd] = pClientRW.WriteFd()
+		attr := &syscall.ProcAttr{
+			Env:   env,
+			Files: files,
 		}
-		l2Cl, err := sources.NewL2Client(l2RPC, logger, nil, l2ClCfg)
+		var err error
+		childPID, err = syscall.ForkExec(os.Args[0], os.Args, attr)
```

Comment:
Instead of forking manually we can also use the `os/exec` package `Command` utils to run a sub-process.

```go
cmd := exec.Command(...) 
cmd.ExtraFiles = []*io.File{pClientRWFile, hClientWFile}` (or something like this, the RW file may require some changes)
err := cmd.Start()
```

Reply:
Done.